src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,5071 @@
+
+        public bool TryReadUInt32(Asn1Tag expectedTag, out uint value)
+        {
+            if (TryReadUnsignedInteger(sizeof(uint), expectedTag, UniversalTagNumber.Integer, out ulong ulongValue))
+            {
+                value = (uint)ulongValue;
+                return true;
+            }
+
+            value = 0;
+            return false;
+        }
+
+        public bool TryReadInt64(out long value) =>
+            TryReadInt64(Asn1Tag.Integer, out value);
+        
+        public bool TryReadInt64(Asn1Tag expectedTag, out long value)
+        {
+            return TryReadSignedInteger(sizeof(long), expectedTag, UniversalTagNumber.Integer, out value);
+        }
+
+        public bool TryReadUInt64(out ulong value) =>
+            TryReadUInt64(Asn1Tag.Integer, out value);
+
+        public bool TryReadUInt64(Asn1Tag expectedTag, out ulong value)
+        {
+            return TryReadUnsignedInteger(sizeof(ulong), expectedTag, UniversalTagNumber.Integer, out value);
+        }
+
+        public bool TryReadInt16(out short value) =>
+            TryReadInt16(Asn1Tag.Integer, out value);
+
+        public bool TryReadInt16(Asn1Tag expectedTag, out short value)
+        {
+            if (TryReadSignedInteger(sizeof(short), expectedTag, UniversalTagNumber.Integer, out long longValue))
+            {
+                value = (short)longValue;
+                return true;
+            }
+
+            value = 0;
+            return false;
+        }
+
+        public bool TryReadUInt16(out ushort value) =>
+            TryReadUInt16(Asn1Tag.Integer, out value);
+
+        public bool TryReadUInt16(Asn1Tag expectedTag, out ushort value)
+        {
+            if (TryReadUnsignedInteger(sizeof(ushort), expectedTag, UniversalTagNumber.Integer, out ulong ulongValue))
+            {
+                value = (ushort)ulongValue;
+                return true;
+            }
+
+            value = 0;
+            return false;
+        }
+
+        public bool TryReadInt8(out sbyte value) =>
+            TryReadInt8(Asn1Tag.Integer, out value);
+
+        public bool TryReadInt8(Asn1Tag expectedTag, out sbyte value)
+        {
+            if (TryReadSignedInteger(sizeof(sbyte), expectedTag, UniversalTagNumber.Integer, out long longValue))
+            {
+                value = (sbyte)longValue;
+                return true;
+            }
+
+            value = 0;
+            return false;
+        }
+
+        public bool TryReadUInt8(out byte value) =>
+            TryReadUInt8(Asn1Tag.Integer, out value);
+
+        public bool TryReadUInt8(Asn1Tag expectedTag, out byte value)
+        {
+            if (TryReadUnsignedInteger(sizeof(byte), expectedTag, UniversalTagNumber.Integer, out ulong ulongValue))
+            {
+                value = (byte)ulongValue;
+                return true;
+            }
+
+            value = 0;
+            return false;
+        }
+
+        private static bool TryCopyPrimitiveBitStringValue(
```

Comment:
Please add comments on this method including parameters

Reply:
The BitString processing got significantly revamped due to comments here and during the over-the-shoulder review.  Hopefully it's a lot clearer now.